M3-5193: RQ-ify tags + enhancements

hkhalil-akamai opened the pull request:
## Description 📝

RQ-ifies all components that hit the `getTags` API endpoint, with some UI and code adjustments:

`TagCell`
* Add loading spinner when adding and deleting tags

`TagDrawer`
* Simplified using the newer `TagsPanel` component

`TagsPanel`
* Fix a bug causing error messages to not display
* Use inline errors rather than toast

`Tags`
* Fix a small style issue when hovering over tags

## Preview 📷

### UI Changes:

<details>
<summary><code>TagCell</code> and <code>TagsPanel</code></summary>

https://user-images.githubusercontent.com/122488130/216154340-3bd91237-24a8-42e5-9dea-b619f1cee791.mov

</details>


<details>
<summary><code>Tags</code></summary>

<details>
<summary>Before</summary>

https://user-images.githubusercontent.com/122488130/216154475-69655671-746d-4d78-900f-9627ac79cd0b.mov

</details>

<details>
<summary>After</summary>

https://user-images.githubusercontent.com/122488130/216154545-97a70ac1-5438-408b-aec2-050a7152fd70.mov

</details>

</details>

## How to test 🧪

* Ensure user tag suggestions work as expected (e.g., in Linodes/NodeBalancers/Kubernetes, etc.)
* Ensure UX of adding and removing tags is clear (e.g., progress indication, errors)

dwiley-akamai reviewed (commented):
I'm going to test the functionality and comb through the code in depth tomorrow, but I did notice that on this branch, if you don't first click the "Add a tag" button on the Linode Detail page and instead go straight into the TagDrawer, the dropdown when you click "Add a tag" in the drawer is not populated with any existing tags (in contrast to Prod, where the dropdown is populated).

hkhalil-akamai commented on packages/manager/src/components/Tag/Tag.tsx:
```diff
@@ -96,7 +102,7 @@ const useStyles = makeStyles((theme: Theme) => ({
 
 export interface Props extends ChipProps {
   label: string;
-  colorVariant?: Variants;
```
All existing usages of this component provide this prop

hkhalil-akamai commented on packages/manager/src/components/TagCell/TagCell.tsx:
```diff
@@ -137,69 +145,18 @@ export const TagCell: React.FC<Props> = (props) => {
     [tags]
   );
 
-  return inTableContext ? (
```
Proposing to delete this: no existing usages of TagCell are in table context

hkhalil-akamai commented on packages/manager/src/components/TagCell/TagDrawer.tsx:
```diff
@@ -33,48 +22,11 @@ export interface TagDrawerProps {
-      <div className={classes.addTag}>
-        <AddTag tags={tags} addTag={_addTag} label={'Add a tag'} />
-      </div>
+      <TagsPanel tags={tags} updateTags={updateTags}></TagsPanel>
```
TagsPanel is a newer component that offers goodies like progress indication

dwiley-akamai replied:
since there are no child components:

```suggestion
<TagsPanel tags={tags} updateTags={updateTags} />
```

hkhalil-akamai commented on packages/manager/src/components/TagsPanel/TagsPanel.tsx:
```diff
@@ -134,56 +133,49 @@ export interface Props {
+    data: userTags,
+    isLoading: userTagsLoading,
+    error: userTagsError,
+  } = useTags(!isRestrictedUser());
```
What's the difference between `isRestrictedUser` (from `permissionsHelper.ts`) and `_isRestrictedUser` (from `useAccountManagement` hook). If they are duplicated, should they be combined?

bnussman-akamai replied:
In my opinion the best way to actually know if a user is restricted is to do the following. I like this way because it is the most direct and 1:1 with the API which aids maintainability. 

```typescript
  const { data: profile } = useProfile();
  const isRestricted = profile?.restricted;
```

Lets avoid using `isRestrictedUser()` when possible because it will not react to changes in the React Query store if we were to mutate it for some reason. `isRestrictedUser()` was created in case we needed to know if a user was restricted in a clunky class component.

The `useAccountManagement` hook is fine. It just uses `useProfile` but has a lot of extra stuff going on.

hkhalil-akamai commented on packages/manager/src/components/TagsPanel/TagsPanel.tsx:
```diff
@@ -195,21 +187,21 @@ const TagsPanel: React.FC<Props> = (props) => {
-          variant: 'error',
-        });
-        setLoading(false);
+      .catch((e) => {
```
Moving this to an inline error because other errors on this page (e.g. duplicate tag, invalid tag, error adding tag) are inline.

hkhalil-akamai commented on packages/manager/src/features/linodes/LinodeEntityDetail.tsx:
```diff
@@ -847,31 +847,18 @@ export const Footer: React.FC<FooterProps> = React.memo((props) => {
-    (tag: string) => {
-      const newTags = linodeTags.filter((thisTag) => thisTag !== tag);
-      updateLinode({ linodeId, tags: newTags }).catch((e) =>
+  const updateTags = React.useCallback(
```
Merging `addTag` and `deleteTag` into `updateTag` to match conventions in other components (e.g., TagsDrawer). However, as a consequence, the error message becomes less specific ("Error updating tag" instead of "Error adding/deleting tag")

dwiley-akamai commented on packages/manager/src/components/TagCell/AddTag.tsx:
```diff
@@ -37,41 +37,35 @@ export type CombinedProps = Props;
   );
+  // @todo should we toast for this? If we swallow the error the only
+  // thing we lose is preexisting tabs as options; the add tag flow
+  // should still work.
```
Outdated comment since errors are being displayed as notices and not toasts now

hkhalil-akamai replied:
This todo refers to the TagCell component (for example, in the Linode details header), which does not have inline errors.

dwiley-akamai commented on packages/manager/src/components/TagCell/TagCell.tsx:
```diff
@@ -210,11 +167,16 @@ export const TagCell: React.FC<Props> = (props) => {
       alignItems="center"
       wrap="nowrap"
     >
+      {loading && (
```
optional: we can use a ternary here instead, i.e. `loading ? ( ... ) : null`

A few years ago we started shifting away from `&&` (based on [this post](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx) IIRC). `loading` is a boolean so we shouldn't ever run into any issues in this specific situation, but personally I just stick with ternaries now.

dwiley-akamai commented on packages/manager/src/components/TagsPanel/TagsPanel.tsx:
```diff
@@ -195,21 +187,21 @@ const TagsPanel: React.FC<Props> = (props) => {
+        setTagError(tagError);
+      })
+      .finally(() => {
+        setTagsLoading(false);
```
What's the argument for not including this in the `catch()` block instead? (same question for the other `.finally()`ies

dwiley-akamai reviewed (commented):
This is looking good overall, nice job! I left a few comments.

This interaction does seem a bit odd:

https://user-images.githubusercontent.com/114682940/216455431-598d0e86-f047-4219-b6c8-84a2c41d5163.mov

This is if you block requests to `*tags*` and go to a NodeBalancer Detail page. The error shows up initially, but then disappears once you click "Add a tag" -- it doesn't come back and the dropdown is just empty.

hkhalil-akamai commented:
> This is looking good overall, nice job! I left a few comments.
> 
> This interaction does seem a bit odd:
> 
>  Screen.Recording.2023-02-02.at.3.51.00.PM.mov 
> This is if you block requests to `*tags*` and go to a NodeBalancer Detail page. The error shows up initially, but then disappears once you click "Add a tag" -- it doesn't come back and the dropdown is just empty.

Good point. Possible solutions:

1. Persisting the error and adding a second line if more errors occur
2. Persisting the latest error on a single line
3. Showing the error in the dropdown instead of "No results"

dwiley-akamai commented on packages/manager/src/queries/tags.ts:
```diff
@@ -0,0 +1,22 @@
+    getTags({ ...params, ...passedParams }, { ...filter, ...passedFilter })
+  )().then((data) => data.data);
+
+export const updateTagsData = (newData: Tag[]): void => {
```
Seems like this one doesn't get used anywhere currently.

I'm thinking this should make use of `useMutation` and we should be able to do some refactoring to use `updateTagsData` instead of `updateTags()`

hkhalil-akamai replied:
The naming in the PR is a bit confusing because this hook gets all the tag suggestions, which there's no API endpoint to update. My understanding is that the API collects all the tags from existing linodes, DBs, etc and returns them as the tag suggestions.

Maybe the hook should be renamed to `useTagSuggestions`?

hkhalil-akamai replied:
But yeah, looks like updateTagsData isn't used anywhere so it should probably be deleted

dwiley-akamai reviewed (commented):
> Good point. Possible solutions:
> 
> 1. Persisting the error and adding a second line if more errors occur
> 2. Persisting the latest error on a single line
> 3. Showing the error in the dropdown instead of "No results"

I would be in favor of persisting the latest error, if one exists, on a single line.

bnussman-akamai commented:
Everything is looking good to me. One thing I see is that whenever we focus the TagsPanel input, tags are fetched. 

Is there a way we can just fetch all tags initially, then from then on, just rely on manual cache mutations to keep RQ up to date?

hkhalil-akamai commented:
> Everything is looking good to me. One thing I see is that whenever we focus the TagsPanel input, tags are fetched.
> 
> Is there a way we can just fetch all tags initially, then from then on, just rely on manual cache mutations to keep RQ up to date?

I updated the Tags query to use the `longLived` preset which significantly cuts back on the number of requests to the tags endpoint.

I also added mutations to roughly keep the suggestions up to date, but I'm not entirely sure how the API synthesizes the list of tag suggestions, but I think appending new tags to the list is a reasonable approximation.

bnussman-akamai reviewed (approved):
Great work on this!

bnussman-akamai commented on packages/manager/src/queries/tags.ts:
```diff
@@ -0,0 +1,25 @@
+
+export const updateTagsSuggestionsData = (newData: Tag[]): void => {
+  const uniqueTags = Array.from(new Set(newData.map((tag) => tag.label)))
+    .sort()
```
Do we need the sort? What function does it serve?

hkhalil-akamai replied:
The API returns tag suggestions sorted alphabetically, so this replicates that behavior. The use of a set is to avoid duplicate suggestions in the list.

bnussman-akamai replied:
> API returns tag suggestions sorted alphabetically

Ah, nice. I didn't know this! Great job!

dwiley-akamai commented on packages/manager/src/components/TagsPanel/TagsPanel.tsx:
```diff
@@ -326,8 +300,10 @@ const TagsPanel: React.FC<Props> = (props) => {
+        {(tagError.length > 0 || userTagsErrorDisplay.length > 0) && (
+          <Typography className={classes.errorNotice}>
+            {tagError.length > 0 ? tagError : userTagsErrorDisplay}
+          </Typography>
```
Instead of wrapping it in `<Typography />`, putting it in a `<Notice error text={...} />` might be more in line with how we present these visually in other areas of the app

hkhalil-akamai replied:
In this case, the `<Notice>` doesn't seem to fit well visually:

Before:
![Screen Shot 2023-02-17 at 5 08 58 PM](https://user-images.githubusercontent.com/122488130/219803238-9a710ee8-9717-4421-aec6-97a4a0d87f0a.jpg)

After:
![image](https://user-images.githubusercontent.com/122488130/219803376-62b72660-58b9-4334-b94a-4af72943e808.png)

dwiley-akamai reviewed (approved):
LGTM, left one minor suggestion of something to look into
